### ISSUE TITLE
The error of expectation is Invalid CA certificate,but the actual error is Unsupported certificate purpose

### ISSUE THREAD

**Issue by 1185129439** [openssl/openssl#14670]
Hello, I've done a test.
What I expect error was that "Invalid CA certificate",but the actual result is 'Unsupported certificate purpose'. 
I think what caused this result is the following code in function check_chain_extensions:
![image](https://user-images.githubusercontent.com/69578444/112241597-867fe600-8c85-11eb-824f-01c29d40caf4.png)

When an error occurs :ctx->error = X509_V_ERR_INVALID_CA, why not return 0? Instead, the value 'ret' may be  refresh when 'ret = check_curve(x);'.
As a matter of fact, it did happen in my tests: the ret is 1.

The client code ：
```
	StartTestCase("VPP_CPP_Req_Test_Peer_V3_Check_001", "v1-v3-v3 and X509_V_FLAG_X509_STRICT");
	Context = SSL_CTX_new(TLS_client_method());
	SSL_CTX_set_options(Context, SSL_OP_NO_TLSv1_3);
	ret = SSL_CTX_set_cipher_list(Context, RSA_AES_128_GCM_SHA256);
	SSL_CTX_set_verify(Context, SSL_VERIFY_PEER, NULL);
	SSL_CTX_set_verify_depth(Context, 20);
	ret = SSL_CTX_load_verify_locations(Context, TESTDATA_PATH"/TLS12_Certs/CertChains/v1-v3-v3/rootcert.pem", NULL);
	
	ssl = SSL_new(Context);
	iSd = create_socket();
	ret = SSL_set_fd_ex(ssl, iSd);	

	param = SSL_get0_param(ssl);
	X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_X509_STRICT);
	
	ret = SSL_connect(ssl);
	printf("\nret: %d, SSL_getLastError: %d, SSL_get_verify_result: %d\n", ret, iLastErr=ERR_GET_REASON(ERR_peek_last_error()), iVerRes=SSL_get_verify_result(ssl));
	TEST_ASSERT((ret != 1) && (iLastErr == 134) && (iVerRes == X509_V_ERR_INVALID_CA),"SSL_connect");
	SSL_free(ssl);
	SSL_CTX_free(Context);
```
the server code：
```
	Context = SSL_CTX_new(TLS_server_method());
	SSL_CTX_set_options(Context, SSL_OP_NO_TLSv1_3);
	ret = SSL_CTX_set_cipher_list(Context, RSA_AES_128_GCM_SHA256);
	SSL_CTX_set_verify_depth(Context, 20);	
	SSL_CTX_set_verify(Context, SSL_VERIFY_NONE, SEC_NULL);
	ret = SSL_CTX_use_certificate_file(Context, TESTDATA_PATH"/TLS12_Certs/CertChains/v1-v3-v3/eecert.der", SSL_FILETYPE_ASN1);
	ret = SSL_CTX_use_PrivateKey_file(Context, TESTDATA_PATH"/TLS12_Certs/CertChains/v1-v3-v3/eekey.der", SSL_FILETYPE_ASN1);
	ret = SSL_CTX_load_verify_locations(Context, TESTDATA_PATH"/TLS12_Certs/CertChains/v1-v3-v3/cacert.pem", NULL);	

	ssl = SSL_new(Context);
	new_sd = create_socket();
	iSd = (int)accept ((unsigned int)new_sd, (struct sockaddr*)&sa_cli, &client_len); 		
	ret = SSL_set_fd_ex(ssl, iSd);	
	ret = SSL_accept (ssl);
	printf("\nret: %d, SSL_getLastError: %d\n", ret, iLastErr=ERR_GET_REASON(ERR_peek_last_error()));	
	SSL_free(ssl);
	SSL_CTX_free(Context);
```
All certificate have been included in the attached zip archive.
[cert.zip](https://github.com/openssl/openssl/files/6193818/cert.zip)

**Comment by kaduk** [comment]
This issue is CVE-2021-3450 and has been resolved in the 1.1.1k security release.

**Comment by richsalz** [comment]
Just to add a little more clarity.  See https://github.com/openssl/openssl/pull/12909 for some background. The project had already been told of this issue, realized it was a CVE and was planning on making a security release before you (@1185129439 ) reported this issue. Because of the security implications, I believe OpenSSL didn't comment on this issue before the release was made public.
